lib/consensus/raft/consensus.go

```diff
@@ -0,0 +1,618 @@
+func (cc *Consensus) WaitForSync(ctx context.Context) error {
+	//ctx, span := trace.StartSpan(ctx, "consensus/WaitForSync")
+	//defer span.End()
+
+	leaderCtx, cancel := context.WithTimeout(ctx, cc.config.WaitForLeaderTimeout)
+	defer cancel()
+
+	// 1 - wait for leader
+	// 2 - wait until we are a Voter
+	// 3 - wait until last index is applied
+
+	// From raft docs:
+
+	// once a staging server receives enough log entries to be sufficiently
+	// caught up to the leader's log, the leader will invoke a  membership
+	// change to change the Staging server to a Voter
+
+	// Thus, waiting to be a Voter is a guarantee that we have a reasonable
+	// up to date state. Otherwise, we might return too early (see
+	// https://github.com/ipfs-cluster/ipfs-cluster/issues/378)
+
+	_, err := cc.raft.WaitForLeader(leaderCtx)
+	if err != nil {
+		return errors.New("error waiting for leader: " + err.Error())
+	}
+
+	err = cc.raft.WaitForVoter(ctx)
+	if err != nil {
+		return errors.New("error waiting to become a Voter: " + err.Error())
+	}
+
+	err = cc.raft.WaitForUpdates(ctx)
+	if err != nil {
+		return errors.New("error waiting for consensus updates: " + err.Error())
+	}
+	return nil
+}
+
+// waits until there is a consensus leader and syncs the state
+// to the tracker. If errors happen, this will return and never
+// signal the component as Ready.
+func (cc *Consensus) finishBootstrap() {
+	// wait until we have RPC to perform any actions.
+	select {
+	case <-cc.ctx.Done():
+		return
+	case <-cc.rpcReady:
+	}
+
+	// Sometimes bootstrap is a no-Op. It only applies when
+	// no state exists and staging=false.
+	_, err := cc.raft.Bootstrap()
+	if err != nil {
+		return
+	}
+
+	logger.Debugf("Bootstrap finished")
+	err = cc.WaitForSync(cc.ctx)
+	if err != nil {
+		return
+	}
+	logger.Debug("Raft state is now up to date")
+	logger.Debug("consensus ready")
+	cc.readyCh <- struct{}{}
+}
+
+// Shutdown stops the component so it will not process any
+// more updates. The underlying consensus is permanently
+// shutdown, along with the libp2p transport.
+func (cc *Consensus) Shutdown(ctx context.Context) error {
+	//ctx, span := trace.StartSpan(ctx, "consensus/Shutdown")
+	//defer span.End()
+
+	//cc.shutdownLock.Lock()
+	//defer cc.shutdownLock.Unlock()
+
+	//if cc.shutdown {
+	//	logger.Debug("already shutdown")
+	//	return nil
+	//}
+
+	logger.Info("stopping Consensus component")
+
+	// Raft Shutdown
+	err := cc.raft.Shutdown(ctx)
+	if err != nil {
+		logger.Error(err)
+	}
+
+	if cc.config.HostShutdown {
+		err = cc.host.Close()
+		if err != nil {
+			logger.Error(err)
+		}
+	}
+
+	//cc.shutdown = true
+	cc.cancel()
+	close(cc.rpcReady)
+	return nil
+}
+
+// Ready returns a channel which is signaled when the Consensus
+// algorithm has finished bootstrapping and is ready to use
+func (cc *Consensus) Ready(ctx context.Context) <-chan struct{} {
+	//_, span := trace.StartSpan(ctx, "consensus/Ready")
+	//defer span.End()
```

Comment:
I'd either drop tracing or uncomment tracing. (we use `"go.opencensus.io/trace"` in lotus, with span names separated by `.`s. Also the namespace should be `raft` as I'm pretty sure we already generate consensus spans, so this would be `trace.StartSpan(ctx, "raft.Ready")`)

Reply:
removed tracing